src/filter.rs

```diff
@@ -190,48 +242,208 @@ pub(crate) fn unfilter(
+                        .zip(previous[8..].chunks_exact(8))
+                    {
+                        let new_chunk = [
+                            chunk[0].wrapping_add(((above[0] as u16 + lprev[0] as u16) / 2) as u8),
+                            chunk[1].wrapping_add(((above[1] as u16 + lprev[1] as u16) / 2) as u8),
+                            chunk[2].wrapping_add(((above[2] as u16 + lprev[2] as u16) / 2) as u8),
+                            chunk[3].wrapping_add(((above[3] as u16 + lprev[3] as u16) / 2) as u8),
+                            chunk[4].wrapping_add(((above[4] as u16 + lprev[4] as u16) / 2) as u8),
+                            chunk[5].wrapping_add(((above[5] as u16 + lprev[5] as u16) / 2) as u8),
+                            chunk[6].wrapping_add(((above[6] as u16 + lprev[6] as u16) / 2) as u8),
+                            chunk[7].wrapping_add(((above[7] as u16 + lprev[7] as u16) / 2) as u8),
+                        ];
+                        *TryInto::<&mut [u8; 8]>::try_into(chunk).unwrap() = new_chunk;
+                        lprev = new_chunk;
+                    }
+                }
             }
-
-            Ok(())
         }
         Paeth => {
-            let previous = require_length(previous, len)?;
-            if bpp > len {
-                return Err("Filtering failed: bytes per pixel is greater than length of row");
-            }
-
-            for i in 0..bpp {
-                current[i] = current[i].wrapping_add(filter_paeth_decode(0, previous[i], 0));
-            }
-
             // Paeth filter pixels:
             // C B D
             // A X
-            for (((i, a_index), &b_pixel), &c_pixel) in
-                (bpp..).zip(0..).zip(&previous[bpp..]).zip(previous)
-            {
-                current[i] = current[i].wrapping_add(filter_paeth_decode(
-                    current[a_index],
-                    b_pixel,
-                    c_pixel,
-                ));
-            }
+            match tbpp {
+                BytesPerPixel::Three => {
+                    assert!(len > 2);
+                    let mut a_bpp = [0; 3];
+                    let mut c_bpp = [0; 3];
+                    for (chunk, b_bpp) in current.chunks_exact_mut(3).zip(previous.chunks_exact(3))
+                    {
+                        let new_chunk = [
+                            chunk[0]
+                                .wrapping_add(filter_paeth_decode(a_bpp[0], b_bpp[0], c_bpp[0])),
+                            chunk[1]
+                                .wrapping_add(filter_paeth_decode(a_bpp[1], b_bpp[1], c_bpp[1])),
+                            chunk[2]
+                                .wrapping_add(filter_paeth_decode(a_bpp[2], b_bpp[2], c_bpp[2])),
+                        ];
+                        *TryInto::<&mut [u8; 3]>::try_into(chunk).unwrap() = new_chunk;
+                        a_bpp = new_chunk;
+                        c_bpp = TryInto::<[u8; 3]>::try_into(b_bpp).unwrap();
```

Comment:
I think the type inference should work if you do `c_bpp = b_bpp.try_into().unwrap()`.